DRAFT: Added /learning-identity section to docs, plus JWT page (#1161)
DuendeSoftware/docs.duendesoftware.com: astro/src/content/docs/learning-identity/what-is-a-jwt.mdx

---
title: What is a JWT
description: Overview of what a JWT is and it's structure.
sidebar:
  label: What is a JWT
  order: 100
---

JSON Web Tokens (JWTs) are JSON formatted strings which contain information about a subject, a human user or some other system like an API service, or AI Agent. The JWT is generated after the subject authenticates with the system by entering their username/password, providing a passkey, or authenticating any other way. 

The official standard for JWTs is the IETF's [RFC7519](https://datatracker.ietf.org/doc/html/rfc7519), and we will be referencing it below.

The structure of a JWT follows the JSON Web Signature (JWS) format defined in [RFC7515](https://datatracker.ietf.org/doc/html/rfc7515). This means they are formatted into three sections separated by a dot. Each section is defined below, but they are all self-contained, Base64 URL Encoded strings.

> > Reminder: Encoded does not mean encrypted. Anyone can decode a JWT and read the plaintext values. The JWT is Base64 URL Encoded to make it simple to transmit the string between services by avoiding characters that can be problematic on the web, like spaces and quotes.

Reviewer: Astro/Starlight has a way to make this a "note", please see other pages for examples.

Author: Will update this. I copy/pasted that from somewhere. I think we need to change a few places to match the correct way.

## Reading JWTs

> Libraries exist to parse a JWT in a variety of programming languages, so you don't need to implement the logic yourself. If you want to make a JWT string human readable so you can view it, you can use any JWT Decoder. Like the one at [Duende's demo site](https://demo.duendesoftware.com/jwt-decoder).

Reviewer: ```suggestion
All programming ecosystems have libraries to parse a JWT, so you don't need to implement the parsing logic yourself, regardless of your technology stack. For the curious, if you want to make your JWT instance human-readable, you can use the [Duende JWT Decoder](https://demo.duendesoftware.com/jwt-decoder).
```

## Sections of a JWT

> The three sections of the JWT are Base64 URL Encoded strings separated by a dot. They are the Header, Payload, and Signature.

Reviewer: ```suggestion
The three sections of the JWT are Base64 URL-encoded strings separated by a dot. They are the **Header**, **Payload**, and **Signature**.
```

An example JWT value `eyJ0eXAiOiJKV1QiLA0KICJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJqb2UiLA0KICJleHAiOjEzMDA4MTkzODAsDQogImh0dHA6Ly9leGFtcGxlLmNvbS9pc19yb290Ijp0cnVlfQ.dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk` would be decoded as:
- Header: `{"typ": "JWT", "alg": "HS256"}`
- Payload: `{ "iss": "joe", "exp": 1300819380, "http://example.com/is_root": true }`
- Signature: `dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk`

### Header

> The header, also called the JSON Object Signing and Encryption Header (JOSE Header), contains properties describing what cryptographic operations were performed to sign the Payload of the JWT. The example below shows that the HS256 algorithm was used to sign the Payload.

Reviewer: ```suggestion
The header, also called the JSON Object Signing and Encryption Header (JOSE Header), contains properties describing the cryptographic operations  performed to sign the JWT payload. The example below shows that the `HS256` algorithm was used to sign the payload.
```

```json
{
  "alg": "HS256",
  "typ": "JWT"
}
```

### Payload

The payload section of the JWT contains the Claims for the subject, which is information about the identity. This includes Registered Claim Names, a set of claims with known meanings, and/or any number of custom claims that are specific to your application(s).

#### Registered Claim Names

Some examples of Registered Claim Names are:
- `sub`: The subject the JWT is for. This can be any string that uniquely identifies the subject, like an email address or GUID.
- `iss`: The issuer of the JWT. This is typically the URI of the system that generated the JWT.
- `iat`: The date/time the JWT was issued. The value is the number of seconds since January 1, 1970. 
- `exp`: The date/time the JWT expires. Systems receiving a JWT should not trust one after its expiration time has passed. The value is the number of seconds since January 1, 1970. 

> The full set of Registered Claim Names are listed in the [JWT RFC](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1).

Reviewer: ```suggestion
The full set of Registered Claim Names is listed in the [JWT RFC](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1).
```

Author: @khalidabuhakmeh Are you sure the suggested change is correct pronunciation? If you are, I'll change it. Just sounds weird to me.

#### Custom Claims

Custom Claims are any claims that are specific to your application(s). For example, you can define the `department` claim to be a string that specifies which department the user works for. The actual value is stored in a database and loaded when the user signs in.

> > Note: An identity system can add any number of claims. Some systems will add everything it knows about the subject, others will scope them only to the system that will use the JWT.

Reviewer: This can be turned into a call out component

#### Real World Example

The below example shows a payload with a mix of claims. The `sub` and `exp` are registered claims, and the rest are custom claims used by the consuming application.

```json
{
  "sub": "user@example.com",
  "exp": 1300819380,
  "client_id": "my-frontend",
  "department": "I.T.",
  "team_name": "core developers",
  "scope": [
    "admin",
    "orders:read",
    "inventory:read",
    "inventory:write"
  ]
}
```

> > Note: A JWT may contain every claim for a user, making the transmitted string very large. Conversely it can contain a minimum amount of information about the subject. The amount of information contained in the JWT is determined by the service generating it.

Reviewer: again, call out component.

### Signature

The signature is used to verify the JWT has not been tampered with since it was created. Because the token is an arbitrary string, the Signature adds trust that the JWT is the same value that it was when it was generated by the identity system (the trusted source of truth). Without verifying the signature, anyone could modify a JWT and get full access to the system.

## Pseudocode Generating a JWT

> If it helps to visualize the process of generating a JWT with code, the below pseudocode describes what steps are taken to generate one:

Reviewer: ```suggestion
If it helps to visualize the process of generating a JWT with code, the following pseudocode describes what steps are taken to generate one:
```
